test/integration/im_cloudbuild_workspace_gitlab/im_cloudbuild_workspace_gitlab_test.go

```diff
@@ -0,0 +1,331 @@
+				mr := client.GetOpenMergeRequest(branch)
+				if mr != nil {
+					client.CloseMergeRequest(mr)
+				}
+				mergeRequest = client.CreateMergeRequest("preview PR", branch, "main")
+				lastCommit = git.GetLatestCommit()
+			case "main":
+				client.AcceptMergeRequest(mergeRequest, "commit message")
+				gitRun("pull")
+				lastCommit = git.GetLatestCommit()
+			}
+
+			// filter builds triggered based on pushed commit sha
+			buildListCmd := fmt.Sprintf("builds list --filter substitutions.COMMIT_SHA='%s' --project %s --region %s --limit 1", lastCommit, projectID, triggerLocation)
+			// poll build until complete
+			pollCloudBuild := func(cmd string) func() (bool, error) {
+				return func() (bool, error) {
+					build := gcloud.Run(t, cmd, gcloud.WithLogger(logger.Discard)).Array()
+					if len(build) < 1 {
+						return true, nil
+					}
+
+					latestWorkflowRunStatus := build[0].Get("status").String()
+					if latestWorkflowRunStatus == "SUCCESS" {
+						t.Logf("josephdthomas - logging the build at end of polling")
+						t.Logf("%v", build)
+						return false, nil
+					}
+					if latestWorkflowRunStatus == "TIMEOUT" || latestWorkflowRunStatus == "FAILURE" {
+						t.Logf("%v", build[0])
+						t.Fatalf("workflow %s failed with status %s", build[0].Get("id"), latestWorkflowRunStatus)
+						return false, nil
+					}
+					return true, nil
+				}
+			}
+			utils.Poll(t, pollCloudBuild(buildListCmd), 20, 15*time.Second)
+			build := gcloud.Run(t, buildListCmd, gcloud.WithLogger(logger.Discard)).Array()[0]
+
+			switch branch {
+			case "preview":
+				assert.Equal(previewTrigger, build.Get("buildTriggerId").String(), "was triggered by preview trigger")
+			case "main":
+				assert.Equal(applyTrigger, build.Get("buildTriggerId").String(), "was triggered by apply trigger")
+			}
+		}
+	})
+
+	bpt.DefineTeardown(func(assert *assert.Assertions) {
+		// Guarantee clean up even if the normal gcloud/teardown run into errors
+		t.Cleanup(func() {
+			client.DeleteProject()
+			bpt.DefaultTeardown(assert)
+		})
+		projectID := bpt.GetStringOutput("project_id")
+		gcloud.Runf(t, "infra-manager deployments delete projects/%s/locations/us-central1/deployments/im-example-gitlab-deployment --project %s --quiet", projectID, projectID)
+	})
+
+	bpt.Test()
+}
+
+// lastElem gets the last element in a string separated by sep.
+// Typically used to grab a resource ID from a full resource name.
+func lastElem(name, sep string) string {
+	return strings.Split(name, sep)[len(strings.Split(name, sep))-1]
+}
+
+// getTerraformExample returns the contents of the example main.tf file.
+func getTerraformExample(t *testing.T) []byte {
+	t.Helper()
+	contents, err := os.ReadFile("files/main.tf")
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	return contents
+}
+
+func getSetupRandomString(t *testing.T) string {
+	t.Helper()
+	setup := tft.NewTFBlueprintTest(t)
+	return setup.GetTFSetupStringOutput("random_testing_string")
+}
```

Comment:
IIRC we used these gh tests too - you can move this into a helper package like https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/test/integration/utils/utils.go